tensorflow/compiler/mlir/tools/kernel_gen/transforms/tf_64bit_indexer.cc

```diff
@@ -0,0 +1,206 @@
+#include "mlir/IR/BuiltinAttributes.h"  // from @llvm-project
+#include "mlir/IR/BuiltinOps.h"  // from @llvm-project
+#include "mlir/IR/Location.h"  // from @llvm-project
+#include "mlir/IR/MLIRContext.h"  // from @llvm-project
+#include "mlir/IR/OperationSupport.h"  // from @llvm-project
+#include "mlir/IR/PatternMatch.h"  // from @llvm-project
+#include "mlir/IR/TypeRange.h"  // from @llvm-project
+#include "mlir/Support/LogicalResult.h"  // from @llvm-project
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"  // from @llvm-project
+#include "tensorflow/compiler/mlir/tensorflow/ir/tf_dialect.h"
+#include "tensorflow/compiler/mlir/tools/kernel_gen/ir/tf_framework_ops.h"
+#include "tensorflow/compiler/mlir/tools/kernel_gen/transforms/passes.h"
+#include "tensorflow/compiler/mlir/tools/kernel_gen/transforms/rewriters.h"
+#include "mlir/Transforms/Bufferize.h"  // from @llvm-project
+
+
+namespace mlir {
+namespace kernel_gen {
+namespace transforms {
+namespace {
+
+
+using shape::ShapeOfOp;
+static constexpr StringRef kEmitCInterfaceAttrName = "llvm.emit_c_interface";
+
+bool IsTFOperation(Operation *op) {
+  return op != nullptr &&
+         op->getDialect() ==
+             op->getContext()->getLoadedDialect<TF::TensorFlowDialect>();
+}
+
+struct TFT64BitIndexerPattern : public RewritePattern {
+  explicit TFT64BitIndexerPattern(MLIRContext *ctx, bool index_64bit)
+      : RewritePattern(MatchAnyOpTypeTag(), /*benefit=*/1, ctx) {index_64bit_ = index_64bit;}
+
+  LogicalResult matchAndRewrite(Operation *op,
+                                PatternRewriter &rewriter) const override {
+    if (!IsTFOperation(op) || op->getParentOfType<tf_framework::JITCompileOp>() ||
+          !llvm::isa<FuncOp>(op->getParentOp()))
+    {
+        return failure();
+    }
+    SmallVector<Operation *, 16> cluster;
+    llvm::SmallPtrSet<Value, 16> operand_set, result_set;
+    Operation *it = op;
+    while (IsTFOperation(it)) {
+      // Find results that escape the JIT compile region.
+      for (auto &use : it->getUses()) {
+        if (!llvm::is_contained(cluster, use.getOwner()))
+          result_set.insert(use.get());
+      }
+
+      // Update JIT region operands and results.
+      for (Value v : it->getResults()) operand_set.erase(v);
+      for (Value v : it->getOperands()) operand_set.insert(v);
+      cluster.push_back(it);
+      it = it->getPrevNode();
+    }
+
+    // Introduce order to the operands and results.
+    auto operands = llvm::to_vector<16>(operand_set);
+    auto results = llvm::to_vector<16>(result_set);
+    auto operand_types = llvm::to_vector<16>(
+        llvm::map_range(operands, [](Value v) { return v.getType(); }));
+    auto result_types = llvm::to_vector<16>(
+        llvm::map_range(results, [](Value v) { return v.getType(); }));
+
+    // Create the JIT compile op.
+    auto loc = op->getLoc();
+    Value shape_size_limit = rewriter.create<arith::ConstantIndexOp>(loc, 4294967296);
```

Comment:
Can we extract `4294967296` as a constant and give it a name?

Reply:
addressed.